openerp/addons/base/migrations/7.0.1.3/post-migration.py

```diff
@@ -240,58 +264,107 @@ def create_partner(address_id, vals, defaults):
+                    cr, SUPERUSER_ID, dict_values)
+                partner_store_insert.append(values)
             else:
-                if address['partner_id'] not in partner_found:
-                    # Main partner address
-                    partner_obj.write(
-                        cr, SUPERUSER_ID, address['partner_id'], partner_vals)
-                    partner_found.append(address['partner_id'])
-                    set_address_partner(address['id'], address['partner_id'])
+                if dict_values['partner_id'] not in partner_found:
+                    for f in ['name', 'id', 'type']:
+                        del dict_values[f]
+                    dict_values['id'] = dict_values['partner_id']
+                    del dict_values['partner_id']
+                    values = dict_values
+                    partner_found.append(values['id'])
+                    partner_store_update.append(values)
                 else:
-                    # any following address for an existing partner
-                    partner_vals.update({
-                        'is_company': False,
-                        'parent_id': address['partner_id']})
-                    propagated_values = partner_obj.read(
-                        cr, SUPERUSER_ID, address['partner_id'],
-                        propagate_fields, load="_classic_write")
-                    propagated_values.pop('id')
-                    partner_vals.update(propagated_values)
-                    create_partner(
-                        address['id'], partner_vals, partner_defaults)
-            processed_ids.append(address['id'])
+                    dict_values.update({'is_company': False})
+                    for f in ['name', 'id', 'type', 'partner_id']:
+                        del dict_values[f]
+                    for key in partner_defaults:
+                        if key not in dict_values:
+                            dict_values[key] = partner_defaults[key]
+                    values = partner_obj._add_missing_default_values(
+                        cr, SUPERUSER_ID, dict_values)
+                    partner_store_insert.append(values)
+            processed_ids.append(
+                values['openupgrade_7_migrated_from_address_id'])
+
+        _insert_partners(cr, pool, partner_store_insert)
+        _update_partners(cr, pool, partner_store_update)
+
+        # Update propagate_fields from partner parent
+        # TODO Use propagate_fields = [ 'lang', 'tz', 'customer', 'supplier',]
+        # to generate this query
+        openupgrade.logged_query(
+            cr, "\n"
+            "UPDATE res_partner part "
+            "SET lang = parent.lang, "
+            "tz = parent.tz, "
+            "customer = parent.customer, "
+            "supplier = parent.supplier "
+            "FROM res_partner parent "
+            "WHERE part.parent_id = parent.id "
+            " AND part.parent_id IS NOT NULL")
```

Comment:
We should only do this with partners we just inserted, right? Otherwise, those values will be reset for existing sub-partners from 6.1.
Also consider sorting partners by their nesting nevel before doing the update, this query will not propagate correctly in multi level hierarchies.